app/utils/auth.py

```
@@ -0,0 +1,277 @@
+
+            logger.info(f"Successfully authenticated user {user_claims.sub} with API key {api_key_record.masked_api_key}")
+            return user_claims
+
+        except Exception as e:
+            if isinstance(e, UnauthorizedAccess):
+                raise
+
+            logger.error(f"Error during API key authentication: {str(e)}")
+            raise UnauthorizedAccess(
+                log_message=f"Authentication failed: {str(e)}"
+            )
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)